### PR TITLE
[LLM Runtime] Refine Python API

### PR DESCRIPTION
**Pull request by zhenwei-intel**
## Type of Change

feature

## Description

- [x] unify generate with generate_tokens
- [x] from_pretrained with not_quant and use_cache
- [x] Support stopping_criteria
- [x] Move check of max_new_tokens and eos_token_id to generate
- [x] get threads from envrion or core nums

## Expected Behavior & Potential Risk

```python
from transformers import AutoTokenizer, TextStreamer
from intel_extension_for_transformers.transformers import AutoModelForCausalLM, WeightOnlyQuantConfig
from transformers import StoppingCriteria, StoppingCriteriaList
import torch
from typing import (IO, TYPE_CHECKING, Any, Callable, Dict, Iterable, List,
                    Literal, Optional, Sequence, Tuple, TypeVar, Union)
model_name = "/mnt/disk1/data2/zhenweil/models/llama/Llama-2-7b-chat-hf"  # or local path to model
woq_config = WeightOnlyQuantConfig(compute_dtype="int8", weight_dtype="int4", use_cache=True)
prompt = "[INST] 1+1=? [\INST]"

tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
inputs = tokenizer(prompt, return_tensors="pt").input_ids
streamer = TextStreamer(tokenizer)


class StopOnTokens(StoppingCriteria):
    def __init__(self, min_length: int, start_length: int, stop_token_id: List[int]):
        self.min_length = min_length
        self.start_length = start_length
        self.stop_token_id = stop_token_id
 
    def __call__(
        self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwargs
    ) -> bool:
        if scores is not None:
            if len(scores) > self.min_length:
                for stop_id in self.stop_token_id:
                    if input_ids[0][self.start_length - 1 + len(scores)] == stop_id:
                        return True
        elif input_ids.shape[-1] - self.start_length > self.min_length:
            for stop_id in self.stop_token_id:
                if input_ids[0][input_ids.shape[-1] - 1] == stop_id:
                    return True
        return False

stopping_criteria = StoppingCriteriaList(
    [
        StopOnTokens(
            min_length=max(100 - 20, 0),
            start_length=inputs.shape[1],
            stop_token_id=[tokenizer.eos_token_id],
        )
    ]
)

model = AutoModelForCausalLM.from_pretrained(model_name, quantization_config=woq_config, trust_remote_code=True)
outputs = model.generate(inputs, streamer=streamer, stopping_criteria=stopping_criteria)
```

original outputs:
```
<s> [INST] 1+1=? [\INST]  The answer to 1+1 is 2. everybody knows that.</s>
```

outputs with stopping_criteria:
```
<s> [INST] 1+1=? [\INST]  The answer to 1+1 is 2. everybody knows that.</s><s> everybody knows that.</s><s> everybody knows that.</s><s> everybody knows that.

So, 1+1=2 is a basic arithmetic operation that is widely accepted and used in various contexts. It is a fundamental concept in mathematics and is used in many areas of study, including algebra, geometry, and calculus.

In mathematics, the operation of adding 1 and 1 is a simple and straightforward process. The result of this operation is 2, which is a basic and fundamental concept in mathematics. This operation is used in many areas of study, including algebra, geometry, and calculus.

In algebra, the operation of adding 1 and 1 is used to solve equations and inequalities. For example, if an equation is 2x + 3 = 5, the operation of adding 1 and 1 would result in 3, which is the solution to the equation.

In geometry, the operation of adding 1 and 1 is used to calculate the area and perimeter of shapes. For example, if a shape has an area of 4 and a perimeter of 8, the operation of adding 1 and 1 would result in an area of 5 and a perimeter of 9.

In calculus, the operation of adding 1 and 1 is used to solve problems involving limits and derivatives. For example, if a function has a limit of 2 and a derivative of 3, the operation of adding 1 and 1 would result in a limit of 3 and a derivative of 4.

In conclusion, the operation of adding 1 and 1 is a fundamental concept in mathematics and is used in many areas of study, including algebra, geometry, and calculus. It is a simple and straightforward process that is widely accepted and used in various contexts.</s>
```

## How has this PR been tested?

how to reproduce the test (including hardware information)

## Dependency Change?

any library dependency introduced or removed

### CODE REVIEW

**Review by hshen14 — COMMENTED** [review]
Do you have example to verify whether stopping criteria works as expected?

**Comment by zhenwei-intel** [comment]
> Do you have example to verify whether stopping criteria works as expected?

update in Expected Behavior & Potential Risk

File: intel_extension_for_transformers/llm/runtime/graph/__init__.py [hunk 1]
```diff
@@ -61,44 +61,61 @@ def __import_package(self, model_name):
...
+        if use_cache and os.path.exists(self.bin_file):
+            return
+
         convert_model(model_name, fp32_bin, "f32")
         assert os.path.exists(fp32_bin), "Fail to convert pytorch model"
 
-        # 2. quant model
-        quant_bin = "ne_{}_q.bin".format(model_type)
-        self.module.Model.quant_model(model_path = fp32_bin, out_path = quant_bin, **kwargs)
+        if not_quant:
```
Comment (kevinintel) [inline intel_extension_for_transformers/llm/runtime/graph/__init__.py:89]: leave messages here

File: intel_extension_for_transformers/llm/runtime/graph/__init__.py [hunk 2]
```diff
@@ -109,34 +126,36 @@ def generate(self, input_ids, streamer=None, interactive=False, ignore_prompt=Fa
...
         if not beam_search:
             # TODO support multi batch
             assert input_ids.shape[0] == 1, "Unsupport multi-batch input ids."
+        
         if streamer:
-            if beam_search:
-                print("ERROR, can not use streamer when use beam search for generation!")
-                import sys
-                sys.exit(1)
+            assert input_ids.shape[0] == 1, "Streamer only supports batch size 1."
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/llm/runtime/graph/__init__.py:137]: So we can use `streamer` when doing beam search?

File: intel_extension_for_transformers/llm/runtime/graph/__init__.py [hunk 3]
```diff
@@ -61,44 +61,61 @@ def __import_package(self, model_name):
...
 
-    def init_from_bin(self, model_name, model_path, **kwargs):
+    def init_from_bin(self, model_name, model_path, **generate_kwargs):
         self.__import_package(model_name)
         self.model = self.module.Model()
-        self.model.init_model(model_path, **kwargs)
+        if "threads" not in generate_kwargs:
+            threads = os.getenv("OMP_NUM_THREADS")
+            if threads is None:
+                generate_kwargs["threads"] = os.cpu_count() // 2
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/__init__.py:103]: why use half cpu for thread num,
assuming hybrid thread is enabling ?

**Review by zhentaoyu — COMMENTED** [review]
since there has `no_quant` arg, maybe need to refine our `test_llm_runtime` python UT?

File: intel_extension_for_transformers/llm/runtime/graph/README.md [hunk 4]
```diff
@@ -88,11 +88,25 @@ outputs = model.generate(inputs, streamer=streamer, max_new_tokens=300, ctx_size
...
+| not_quant         | Bool        | Determine whether or not the model will be quantized. (default: False)                  |
+| use_cache         | Bool        | Use local quantized model if file exists (default: False)                               |
+
 Argument description of generate function:
 | Argument          |  Type       | Description                                                                             |
 | --------------    | ----------  | -----------------------------------------------------------------------                 |
 | inputs            | Lists[Int]  | Input ids after tokenizer                                                               |
 | streamer          | Class       | Streamer object that will be used to stream the generated sequences. (default: None)    |
+| stopping_criteria | Class       | Custom stopping criteria that complement the default stopping criteria built from arguments
+ and a generation config. (default: None)    |
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/README.md:109]: maybe we can provide a template for stopping_criteria  usage.

File: intel_extension_for_transformers/llm/runtime/graph/README.md [hunk 5]
```diff
@@ -88,11 +88,25 @@ outputs = model.generate(inputs, streamer=streamer, max_new_tokens=300, ctx_size
...
+| weight_dtype      | String      | Data type of quantized weight: int4/int8 (default int4)                                 |
+| alg               | String      | Quantization algorithm: sym/asym (default sym)                                          |
+| group_size        | Int         | Group size: Int (default: 32)                                                           |
+| scale_dtype       | String      | Data type of scales: fp32/bf16 (dafault fp32)                                           |
+| use_ggml          | Bool        | Enable ggml for quantization and inference (default: False)                             |
+| not_quant         | Bool        | Determine whether or not the model will be quantized. (default: False)                  |
+| use_cache         | Bool        | Use local quantized model if file exists (default: False)                               |
+
 Argument description of generate function:
 | Argument          |  Type       | Description                                                                             |
```
[the remark this answers is not in the document]
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/README.md:104]: updated c0bab8101f41e1fe5f0625dce866fe61911ed6e0

File: intel_extension_for_transformers/llm/runtime/graph/__init__.py (hunk 1 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/__init__.py:89]: update 0fdb9e848fd214fb2770650ec629c049aad1686a

File: intel_extension_for_transformers/llm/runtime/graph/__init__.py (hunk 3 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/__init__.py]: update 0fdb9e848fd214fb2770650ec629c049aad1686a,
Use `len(os.sched_getaffinity(0))` to get cores under numa settings.  thanks @DDEle

**Review by airMeng — APPROVED** [review]

**Review by DDEle — APPROVED** [review]

File: intel_extension_for_transformers/llm/runtime/graph/README.md (hunk 4 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/README.md]: update readme 278aecb2193a0c40d30d145eee1fa38ba299c184

File: intel_extension_for_transformers/llm/runtime/graph/__init__.py (hunk 2 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/__init__.py:138]: cannot, update f71da91bf8c78a6d7376675d10d67a1996c10495

**Review by zhentaoyu — APPROVED** [review]

**Comment by zhenwei-intel** [comment]
> since there has `no_quant` arg, maybe need to refine our `test_llm_runtime` python UT?

update test

**Review by a32543254 — APPROVED** [review]
LGTM
